WIP: sys/linux: add new FUTEX_WAIT_MULTIPLE operation

andrealmeid opened the pull request:
Hello,

This is a WIP, since the futex operation described at this PR hasn't been merged yet, so it can't be found at linux-next tree. I'm being using the tool to fuzz the `FUTEX_WAIT_MULTIPLE` operation and I'm opening this PR in the hope to get some feedback about my changes and eventually merge it. You can check the patch at my public branches[1][2] and at LKML[3][4].

### How this new operation works?

In summary, this operation allows a waiter to wait on multiple futexes at same time, and wake up at the first one available. To describe the futexes to be wait for, one needs to create an array of `struct futex_wait_block`:

```c
struct futex_wait_block {
	int *uaddr;
	int val;
	__u32 bitset;
};
```
The length of the array should be between `[1, 128]`. Then, to wait, just use the futex syscall:

```c
/**
 * futex_wait_multiple() - block on several futexes with optional timeout
 * @fwb:	wait block user space address
 * @count:	number of entities at fwb
 * @timeout:	absolute timeout
 */
int futex_wait_multiple(struct futex_wait_block *fwb, int count,
                        struct timespec *timeout, int opflags)
{
	return syscall(SYS_futex, fwb, FUTEX_WAIT_MULTIPLE, count, timeout, NULL, 0,
		     opflags);
}
```
Details of implementation can be found at commit messages, at cover letter and at the `kselftest` examples.

Thanks!

---

[1] https://gitlab.collabora.com/tonyk/linux/commits/futex-next-v3
[2] https://gitlab.collabora.com/tonyk/linux/commits/futex-dev-v3
[3] https://lkml.org/lkml/2020/2/13/1475
[4] https://lore.kernel.org/patchwork/cover/1194338/

dvyukov commented:
Hi André,

I don't see lots of WIP here, you did everything properly, constants for all arches and everything is regenerated and checked-in. I have few small nits, but otherwise it's pretty merge-able.
What are the current moods upstream wrt merging this eventually?

andrealmeid commented:
> I don't see lots of WIP here, you did everything properly, constants for all arches and everything is regenerated and checked-in. I have few small nits, but otherwise it's pretty merge-able.

Cool, thanks :)

> What are the current moods upstream wrt merging this eventually?

The current version of the patch didn't get much attention from the maintainers. But, since we addressed all feedback and question of the first iteration, we hope it will be merged eventually.

I'll address your feedbacks and update this PR.

andrealmeid commented:
@dvyukov updated the PR, could you have a  look?

dvyukov commented:
Perfect! Thanks!

andrealmeid commented:
Thanks @dvyukov :)
